### PR TITLE
    fix(tianmu): fix query result err when executing UNION has NULL Clause and datetime precision of union (#998)

### LINKED ISSUE #998: bug: Result set Error, when executing UNION Clause
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

the wrong result is as follows:

```shell
mysql> SELECT t1_int FROM t1 UNION SELECT t2_char FROM t2;
+-------------+
| t1_int      |
+-------------+
| -2147483648 |
| 1           |
| 2           |
| 3           |
| 4           |
| 5           |
| 6           |
| 7           |
|             |
| eee         |
| ccc         |
| jjj         |
| lll         |
| nnn         |
+-------------+
14 rows in set (0.01 sec)
```

### Expected behavior

```shell
mysql> SELECT t1_int FROM t1 UNION SELECT t2_char FROM t2;
+--------+
| t1_int |
+--------+
| NULL   |
| 1      |
| 2      |
| 3      |
| 4      |
| 5      |
| 6      |
| 7      |
|        |
| eee    |
| ccc    |
| jjj    |
| lll    |
| nnn    |
+--------+
14 rows in set (0.00 sec)
```

### How To Reproduce

```sql
CREATE TABLE t1 (t1_int INT, t1_char CHAR(5));
CREATE TABLE t2 (t2_int INT, t2_char CHAR(5));
INSERT INTO t1 VALUES (NULL, ''),(1, 'aaa'),(2, 'aaa'),(3, 'ccc'),(4, 'ddd'),(5, 'aaa'),(6, ''),(7, 'eee');
INSERT INTO t2 VALUES (NULL, ''),(1, 'eee'),(3, 'ccc'),(5, 'jjj'),(6, ''),(7, 'lll'),(9, 'eee'),(11, 'nnn');
SELECT t1_int FROM t1 UNION SELECT t2_char FROM t2;
```

### Environment

```shell
[root@HAST04 ~]# /stonedb57/install/bin/mysqld --version
/stonedb57/install/bin/mysqld  Ver 5.7.36-StoneDB for Linux on x86_64 (build-)
build information as follow:
        Repository address: https://github.com/stoneatom/stonedb.git:stonedb-5.7-dev
        Branch name: stonedb-5.7-dev
        Last commit ID: 7f69d33
        Last commit time: Date:   Tue Nov 22 06:31:04 2022 +0000
        Build time: Date: Tue Nov 22 16:53:27 CST 2022

[root@HAST04 ~]# cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
```

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by adofsauron**
Reasons: Null value is not checked

## Summary about this PR

Issue Number: close #998 

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/core/temp_table.cpp [hunk 1]
```diff
@@ -468,28 +468,28 @@ void TempTable::Attr::GetValueString(types::BString &s, int64_t obj) {
       break;
     case common::CT::BYTEINT: {
       types::TianmuNum tianmu_n((int64_t)(*(AttrBuffer<char> *)buffer)[obj], -1, false, TypeName());
```
Comment (RingsC) [inline storage/tianmu/core/temp_table.cpp]: c++ style type casting.

File: storage/tianmu/core/temp_table.cpp [hunk 2]
```diff
@@ -468,28 +468,28 @@ void TempTable::Attr::GetValueString(types::BString &s, int64_t obj) {
       break;
     case common::CT::BYTEINT: {
       types::TianmuNum tianmu_n((int64_t)(*(AttrBuffer<char> *)buffer)[obj], -1, false, TypeName());
-      s = tianmu_n.ToBString();
+      s = tianmu_n.GetValueInt64() == common::NULL_VALUE_64 ? types::BString() : tianmu_n.ToBString();
```
Comment (RingsC) [inline storage/tianmu/core/temp_table.cpp]: Logic is not ok. we should ues the corresponding data type, not always int64

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/temp_table.cpp (hunk 2 above)
Comment (RingsC) [reply storage/tianmu/core/temp_table.cpp]: > Logic is not ok. we should ues the corresponding data type, not always int64

We will refactor it in future.

File: storage/tianmu/core/temp_table.cpp [hunk 3]
```diff
@@ -468,28 +468,28 @@ void TempTable::Attr::GetValueString(types::BString &s, int64_t obj) {
...
     case common::CT::BYTEINT: {
       types::TianmuNum tianmu_n((int64_t)(*(AttrBuffer<char> *)buffer)[obj], -1, false, TypeName());
-      s = tianmu_n.ToBString();
+      s = tianmu_n.GetValueInt64() == common::NULL_VALUE_64 ? types::BString() : tianmu_n.ToBString();
       break;
     }
     case common::CT::SMALLINT: {
       types::TianmuNum tianmu_n((int64_t)(*(AttrBuffer<short> *)buffer)[obj], -1, false, TypeName());
-      s = tianmu_n.ToBString();
+      s = tianmu_n.GetValueInt64() == common::NULL_VALUE_64 ? types::BString() : tianmu_n.ToBString();
```
Comment (RingsC) [inline storage/tianmu/core/temp_table.cpp]: pls, re-naming `s` with meaningful name.

**Review by RingsC — CHANGES_REQUESTED** [review]

**Review by lujiashun — APPROVED** [review]
LGTM

**Review by RingsC — APPROVED** [review]
LGTM

**Pull request merged**